add 'update_password' param to manageiq_user

joelddiaz opened the pull request:
##### SUMMARY
Currently with the manageiq_user module, if you call it repeatedly while passing the 'password' parameter, it will always run the task and mark it as 'changed'.

Following the pattern of the AWS IAM module, add an 'update_password' parameter that takes 'always' (default) or 'on_create'. This will let you set an initial password when creating a user, but allow the user to modify their password and not stomp over their password changes if you re-run the playbook/task that created the user.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
manageiq_user

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel 31d2eb0828) last updated 2017/09/06 15:15:51 (GMT -400)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/jdiaz/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/jdiaz/tmp/ansible/lib/ansible
  executable location = /home/jdiaz/tmp/ansible/bin/ansible
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```
    - name: create user
      manageiq_user:
        state: present
        userid: testuser
        password: supersecret
        name: Test User
        email: testuser@testing123.com
        group: EvmGroup-user
        manageiq_connection: "{{ miq_connection }}"

Before if you called the above task twice the second run would return 'changed':
changed: [127.0.0.1] => {"changed": true, "failed": false, "msg": "successfully updated the user testuser: {u'name': u'Test User', u'settings': {}, u'userid': u'testuser', u'current_group_id': u'4', u'id': u'38', u'created_on': u'2017-09-07T15:37:12Z', u'href': u'https://REDACTED/api/users/38', u'updated_on': u'2017-09-07T15:37:22Z', u'email': u'testuser@testing123.com'}"}

After the change (and adding the 'update_password' param set to 'on_create') running the task twice will show no changes on the second call:

ok: [127.0.0.1] => {"changed": false, "failed": false, "msg": "user testuser is not changed."}
```

Commit:
add 'update_password' param to manageiq_user

Currently with the manageiq_user module, if you call it repeatedly while passing the 'password' parameter, it will always run the task and mark it as 'changed'.

Following the pattern of the AWS IAM module, add an 'update_password' parameter that takes 'always' (default) or 'on_create'. This will let you set an initial password when creating a user, but allow the user to modify their password and not stomp over their password changes if you re-run the playbook/task that created the user.

yaacov reviewed (approved):
Thanks !
LGTM 👍

yaacov commented:
shipit

yaacov commented:
I think they want you to update version from 2.4. to 2.5 in line 35 ?

yaacov commented:
i will re "ship it" once it pass :-)

joelddiaz commented:
@yaacov line 35 updated to 2.5

yaacov commented:
ahhh my bad !

version_added for new option (update_password) should be 2.5
they want the `version_added` not on the module ( line 35 as I thought ) but on the option itself ? line 72

yaacov commented:
yes, see here:
https://github.com/ansible/ansible/blob/d8b1cb9a634d0e828a7be39d4019a6215410e643/lib/ansible/modules/windows/win_uri.py#L78

the version_added in on the field added

joelddiaz commented:
ok, i put it back to 2.4 for the global version, and put the 'version_added' under the 'update_password' parameter

yaacov reviewed (approved):
Thanks !

yaacov commented:
shipit

yaacov commented:
@cben @zgalor please review ( and "ship it" if ready )

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_user.py:
```diff
@@ -183,7 +190,7 @@ def compare_user(self, user, name, group_id, password, email):
         found_difference = (
             (name and user['name'] != name) or
-            (password is not None) or
+            (password is not None and self.module.params['update_password'] == 'always') or
```
Is this enough?  IIUC this allows the module to skip POST if everything but password matches, but if something else is different (say port), it will stomp the password too.
I think you need logic in update to omit password from API request?

yaacov replied:
a.
The issue is: we send update post request each time if "password" is set, because we can not compare it to current value.

This code fixes this issue.

b.
New issue: The name "update_password" implies that we do not update password, yes but, a name like "do_not_cheack_password_changes" is more confusing. current option is "just right" by me.

keep in mind the even is "update_password" is "on_creat", we do want to update the password is it's changed, or something else is changed, we just not want to send it always (see original issue) ...

c. Nice catch, we always update all fields and not just the changed ones, this is not specific to passwords, a code that send only the changed values will be nice (in a new PR)

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_user.py:
```diff
@@ -280,7 +287,9 @@ def main():
+            state=dict(choices=['absent', 'present'], default='present'),
+            update_password=dict(choices=['always', 'on_create'],
+                                 default='always')
         ),
```
Can you add trailing comma to avoid future noise?
(Ignore if ansible's lint disagrees)

cben commented:
Nice! We went back and forth on password handling a few times, hadn't considered this use case, didnt know there is existing pattern in ansible.

biomassives commented:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (104/100 (104%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext2.iad.github.net ([192.30.252.193]:41475 helo=github-smtp2b-ext-cp1-prd.iad.github.net)
	by chi-server32.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89)
	(envelope-from <noreply@github.com>)
	id 1drOIv-004FRJ-AD
	for greg@biomassiv.es; Mon, 11 Sep 2017 07:59:09 -0500
Date: Mon, 11 Sep 2017 05:58:28 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1505134708;
	bh=fWg1Q1vvyxI52LEc1WyGrXAz8LyyneDgd9sMXkWkM8A=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=QNigXzijA9uagGA8ASJSBxPeak78quew4Y71U6GibD6StBpKXuEYRYoJ+BnclXi9N
	 lOdwiD0mSEMprG7p5rCUsBbM8o3bVqJ1x8b7I2T7TMOKR4VmLufCxqCpnbUH991Vv4
	 Am1kwWHm3a/Bxibq1pJ8ZPlqNEIAgZ5aUKG/jz2U=
From: Yaacov Zamir <notifications@github.com>
Reply-To: ansible/ansible <reply@reply.github.com>
To: ansible/ansible <ansible@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Message-ID: <ansible/ansible/pull/29093/review/61808109@github.com>
In-Reply-To: <ansible/ansible/pull/29093@github.com>
References: <ansible/ansible/pull/29093@github.com>
Subject: Re: [ansible/ansible] add 'update_password' param to manageiq_user
 (#29093)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_59b68874e3546_53e33faf664a3c3813081e";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: yaacov
X-GitHub-Recipient: biomassives
X-GitHub-Reason: subscribed
List-ID: ansible/ansible <ansible.ansible.github.com>
List-Archive: https://github.com/ansible/ansible
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+0042d4e2457f3b465bfc248c6011fe9d1bad580724d374bb92cf0000000115ce4a7492a169ce0f420000@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AELU4j_D5-pCasq3PiQ5zfVqy9QWdpzPks5shS50gaJpZM4PQA_A>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: greg@biomassiv.es


----==_mimepart_59b68874e3546_53e33faf664a3c3813081e
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

yaacov commented on this pull request.



> @@ -183,7 +190,7 @@ def compare_user(self, user, name, group_id, password, email):
         """
         found_difference = (
             (name and user['name'] != name) or
-            (password is not None) or
+            (password is not None and self.module.params['update_password'] == 'always') or

a.
The issue is: we send update post request each time if "password" is set, because we can not compare it to current value.

This code fixes this issue.

b.
New issue: The name "update_password" implies that we do not update password, yes but, a name like "do_not_cheack_password_changes" is more confusing. current option is "just right" by me.

keep in mind the even is "update_password" is "on_creat", we do want to update the password is it's changed, or something else is changed, we just not want to send it always (see original issue) ...

c. Nice catch, we always update all fields and not just the changed ones, this is not specific to passwords, a code that send only the changed values will be nice (in a new PR)

-- 
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/ansible/ansible/pull/29093#discussion_r138060934
----==_mimepart_59b68874e3546_53e33faf664a3c3813081e
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

<p><b>@yaacov</b> commented on this pull request.</p>

<hr>

<p>In <a href="https://github.com/ansible/ansible/pull/29093#discussion_r138060934">lib/ansible/modules/remote_management/manageiq/manageiq_user.py</a>:</p>
<pre style='color:#555'>&gt; @@ -183,7 +190,7 @@ def compare_user(self, user, name, group_id, password, email):
         &quot;&quot;&quot;
         found_difference = (
             (name and user[&#39;name&#39;] != name) or
-            (password is not None) or
+            (password is not None and self.module.params[&#39;update_password&#39;] == &#39;always&#39;) or
</pre>
<p>a.<br>
The issue is: we send update post request each time if "password" is set, because we can not compare it to current value.</p>
<p>This code fixes this issue.</p>
<p>b.<br>
New issue: The name "update_password" implies that we do not update password, yes but, a name like "do_not_cheack_password_changes" is more confusing. current option is "just right" by me.</p>
<p>keep in mind the even is "update_password" is "on_creat", we do want to update the password is it's changed, or something else is changed, we just not want to send it always (see original issue) ...</p>
<p>c. Nice catch, we always update all fields and not just the changed ones, this is not specific to passwords, a code that send only the changed values will be nice (in a new PR)</p>

<p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">&mdash;<br />You are receiving this because you are subscribed to this thread.<br />Reply to this email directly, <a href="https://github.com/ansible/ansible/pull/29093#discussion_r138060934">view it on GitHub</a>, or <a href="https://github.com/notifications/unsubscribe-auth/AELU4q6SnfipGFgkjXAX6j4XPvGBrxmWks5shS50gaJpZM4PQA_A">mute the thread</a>.<img alt="" height="1" src="https://github.com/notifications/beacon/AELU4sMekcVSFCG6eY6Gjf6hzTNh4cOuks5shS50gaJpZM4PQA_A.gif" width="1" /></p>
<div itemscope itemtype="http://schema.org/EmailMessage">
<div itemprop="action" itemscope itemtype="http://schema.org/ViewAction">
  <link itemprop="url" href="https://github.com/ansible/ansible/pull/29093#discussion_r138060934"></link>
  <meta itemprop="name" content="View Pull Request"></meta>
</div>
<meta itemprop="description" content="View this Pull Request on GitHub"></meta>
</div>

<script type="application/json" data-scope="inboxmarkup">{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/ansible/ansible","title":"ansible/ansible","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/ansible/ansible"}},"updates":{"snippets":[{"icon":"PERSON","message":"@yaacov commented on #29093"}],"action":{"name":"View Pull Request","url":"https://github.com/ansible/ansible/pull/29093#discussion_r138060934"}}}</script>
----==_mimepart_59b68874e3546_53e33faf664a3c3813081e--

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_user.py:
```diff
@@ -183,7 +190,7 @@ def compare_user(self, user, name, group_id, password, email):
         found_difference = (
             (name and user['name'] != name) or
-            (password is not None) or
+            (password is not None and self.module.params['update_password'] == 'always') or
```
(c) is harmless, except password which we never know old value.

(b) if this is desired behavior, i disagree with keeping this name.  if the use case is "admin uses ansible to create/update user, but don't stomp on user's password", this is not desired behavior.
IMO, if I give module param A, what happens to it should never depend on whether param B changed.

yaacov commented on lib/ansible/modules/remote_management/manageiq/manageiq_user.py:
```diff
@@ -183,7 +190,7 @@ def compare_user(self, user, name, group_id, password, email):
         found_difference = (
             (name and user['name'] != name) or
-            (password is not None) or
+            (password is not None and self.module.params['update_password'] == 'always') or
```
@joelddiaz ^^

a. what is the desired behavior ?
b. do want to re-set the password if the object is changed ?

if we do, @cben suggests to use a different name, from "update_password"  "on_create" it may not be clear the we cosider any chage a "creation". 

I think that the current name "update_password" is the best one, considering the options :-)
@cben can you think of better name ?

yaacov commented on lib/ansible/modules/remote_management/manageiq/manageiq_user.py:
```diff
@@ -183,7 +190,7 @@ def compare_user(self, user, name, group_id, password, email):
         found_difference = (
             (name and user['name'] != name) or
-            (password is not None) or
+            (password is not None and self.module.params['update_password'] == 'always') or
```
@joelddiaz @cben 

how about  "update_password" "on_user_change" ?

joelddiaz replied:
The desired behavior was that I can have a playbook that lays down MIQ users. Later, the users change their passwords. I re-run the 'ensure-users-exists' playbook, and it doesn't lay down the old/original passwords during the re-run.

I don't think we want to re-set the password when other fields change. There would need to be some additional filtering before creating the POST request when update_password==on_create

yaacov commented on lib/ansible/modules/remote_management/manageiq/manageiq_user.py:
```diff
@@ -183,7 +190,7 @@ def compare_user(self, user, name, group_id, password, email):
         found_difference = (
             (name and user['name'] != name) or
-            (password is not None) or
+            (password is not None and self.module.params['update_password'] == 'always') or
```
> The desired behavior was that I can have a playbook that lays down MIQ users. Later, the users change their passwords. I re-run the 'ensure-users-exists' playbook, and it doesn't lay down the old/original passwords during the re-run.

Great ! , then we need to **delete the passwrod** if a user exist, like @cben pointed out, current approach may re-set the password inadvertently  if user changes,
so we do not need the check here, all we need is to delete the password if the user exits and we are "on_create", this check will fail because we will not have password in the params

yaacov replied:
Nice @cben 👍 ( I didn't understand what @joelddiaz intended here ... )

yaacov commented on lib/ansible/modules/remote_management/manageiq/manageiq_user.py:
```diff
@@ -183,7 +190,7 @@ def compare_user(self, user, name, group_id, password, email):
         found_difference = (
             (name and user['name'] != name) or
-            (password is not None) or
+            (password is not None and self.module.params['update_password'] == 'always') or
```
@joelddiaz once you explained it makes perfect seance to me :-) , but I think we need some more verbose documentation or an example ...

Commit:
don't stomp password when other fields change

Handle case where user fields change, but we don't want to stomp on a potentially user-changed password. Previously, if a non-password field changed, and the password param was passed in, it would ignore the 'update_password': 'on_create' setting (ie it would update/modify the password even if the user already exists).

Add trailing ',' to list of params.

joelddiaz commented:
@yaacov @cben I added the logic for the case when non-password fields are detected as changed so that it does not send the old/original password when update_password is set to 'on_create'. Also added the trailing comma as requested.

yaacov commented on lib/ansible/modules/remote_management/manageiq/manageiq_user.py:
```diff
@@ -232,6 +239,8 @@ def edit_user(self, user, name, group, password, email):
 
         # try to update user
         try:
+            if self.module.params['update_password'] == 'on_create':
```
if you do it before the compare, you will not need to change the compare function, password will be none

yaacov replied:
I think the easiest place to add it will be in main before we call `manageiq_user.edit_user` , and make sure password is None before be use the funtion if we are "on_create".
That way, it will be the only change needed ( no changes to any other function needed)
And it will also be clear to someone reading the code, that we ignore the password param on edit when the "update_password" in "on_create"

joelddiaz commented:
@yaacov I removed the duplicated 'on_create' logic and moved it to set things up properly before the comparison/POST actions.

yaacov reviewed (approved):
LGTM 👍

yaacov commented:
shipit

yaacov commented:
@cben PTAL

yaacov commented on lib/ansible/modules/remote_management/manageiq/manageiq_user.py:
```diff
@@ -219,11 +226,16 @@ def edit_user(self, user, name, group, password, email):
 
+        # if there is a password param, but 'update_password' is 'on_create'
+        # then discard the password (since we're editing an existing user)
+        if password is not None and self.module.params['update_password'] == 'on_create':
```
Not a blocker, I like better:

```
if self.module.params['update_password'] == 'on_create':
   password = None
if password is not None:
   resource['password'] = password
```

I doed the exact thing, with less words :-)

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_user.py:
```diff
@@ -219,11 +226,16 @@ def edit_user(self, user, name, group, password, email):
+        # if there is a password param, but 'update_password' is 'on_create'
+        # then discard the password (since we're editing an existing user)
+        if password is not None and self.module.params['update_password'] == 'on_create':
+            password = None
```
this works fine but the if/elif is slightly complicated to follow, one branch touching password the other resource['password']...
Consider something like this:
```
if self.module.params['update_password'] == 'on_create':
  password = None

if password is not None:
   resource['password'] = password
```

joelddiaz commented:
@yaacov @cben I updated the logic as suggested to help make it more clear.

yaacov reviewed (approved):
LGTM 👍

yaacov commented:
> @yaacov @cben I updated the logic as suggested to help make it more clear.

Nice !

yaacov commented:
shipit

yaacov commented:
@zgalor we need your shipit

cben reviewed (approved):
shipit

zgalor reviewed (approved):
shipit
